File: backend/routers/colleges.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from fastapi import APIRouter, HTTPException, Query
from typing import Optional, List
from pydantic import BaseModel
from database import get_conn, enrich

router = APIRouter(prefix="/api/colleges", tags=["colleges"])


class CollegeIn(BaseModel):
    name: str
    city: str
    type: str
    category: str
    fees: int
    placement: Optional[str] = None
    avg_package: Optional[str] = None
    highest_package: Optional[str] = None
    seats: Optional[int] = None
    criteria: Optional[str] = None
    description: Optional[str] = None
    courses: List[str] = []
    branches: List[str] = []
    facilities: List[str] = []


class CollegeUpdate(BaseModel):
    name: Optional[str] = None
    city: Optional[str] = None
    type: Optional[str] = None
    category: Optional[str] = None
    fees: Optional[int] = None
    placement: Optional[str] = None
    avg_package: Optional[str] = None
    highest_package: Optional[str] = None
    seats: Optional[int] = None
    criteria: Optional[str] = None
    description: Optional[str] = None
    courses: Optional[List[str]] = None
    branches: Optional[List[str]] = None
    facilities: Optional[List[str]] = None


@router.get("/")
def list_colleges(
    city: Optional[str] = Query(None),
    type: Optional[str] = Query(None),
    category: Optional[str] = Query(None),
    maxFees: Optional[int] = Query(None),
    search: Optional[str] = Query(None),
):
    conn = get_conn()
    sql = "SELECT * FROM colleges WHERE 1=1"
    params = []
    if city:
        sql += " AND city = ?"; params.append(city)
    if type:
        sql += " AND type = ?"; params.append(type)
    if category:
        sql += " AND category = ?"; params.append(category)
    if maxFees:
        sql += " AND fees <= ?"; params.append(maxFees)
    if search:
        sql += " AND (name LIKE ? OR city LIKE ? OR description LIKE ?)"
        like = f"%{search}%"
        params.extend([like, like, like])
    rows = [dict(r) for r in conn.execute(sql, params).fetchall()]
    result = [enrich(conn, r) for r in rows]
    conn.close()
    return result


@router.get("/{college_id}")
def get_college(college_id: int):
    conn = get_conn()
    row = conn.execute("SELECT * FROM colleges WHERE id = ?", (college_id,)).fetchone()
    if not row:
        conn.close()
        raise HTTPException(status_code=404, detail="College not found")
    result = enrich(conn, dict(row))
    conn.close()
    return result


@router.post("/", status_code=201)
def create_college(data: CollegeIn):
    conn = get_conn()
    cur = conn.cursor()
    cur.execute("""
        INSERT INTO colleges (name, city, type, category, fees, placement,
            avg_package, highest_package, seats, criteria, description)
        VALUES (?,?,?,?,?,?,?,?,?,?,?)
    """, (data.name, data.city, data.type, data.category, data.fees,
          data.placement, data.avg_package, data.highest_package,
          data.seats, data.criteria, data.description))
    cid = cur.lastrowid
    for c in data.courses:
        cur.execute("INSERT INTO college_courses (college_id, course) VALUES (?,?)", (cid, c))
    for b in data.branches:
        cur.execute("INSERT INTO college_branches (college_id, branch) VALUES (?,?)", (cid, b))
    for f in data.facilities:
        cur.execute("INSERT INTO college_facilities (college_id, facility) VALUES (?,?)", (cid, f))
    conn.commit()
    row = conn.execute("SELECT * FROM colleges WHERE id = ?", (cid,)).fetchone()
    result = enrich(conn, dict(row))
    conn.close()
    return result


@router.put("/{college_id}")
def update_college(college_id: int, data: CollegeUpdate):
    conn = get_conn()
    if not conn.execute("SELECT id FROM colleges WHERE id = ?", (college_id,)).fetchone():
        conn.close()
        raise HTTPException(status_code=404, detail="College not found")
    fields = {k: v for k, v in data.model_dump().items()
              if v is not None and k not in ("courses", "branches", "facilities")}
    if fields:
        set_clause = ", ".join(f"{k} = ?" for k in fields)
        conn.execute(f"UPDATE colleges SET {set_clause} WHERE id = ?",
                     list(fields.values()) + [college_id])
    if data.courses is not None:
        conn.execute("DELETE FROM college_courses WHERE college_id = ?", (college_id,))
        for c in data.courses:
            conn.execute("INSERT INTO college_courses (college_id, course) VALUES (?,?)", (college_id, c))
    if data.branches is not None:
        conn.execute("DELETE FROM college_branches WHERE college_id = ?", (college_id,))
        for b in data.branches:
            conn.execute("INSERT INTO college_branches (college_id, branch) VALUES (?,?)", (college_id, b))
    if data.facilities is not None:
        conn.execute("DELETE FROM college_facilities WHERE college_id = ?", (college_id,))
        for f in data.facilities:
            conn.execute("INSERT INTO college_facilities (college_id, facility) VALUES (?,?)", (college_id, f))
    conn.commit()
    row = conn.execute("SELECT * FROM colleges WHERE id = ?", (college_id,)).fetchone()
    result = enrich(conn, dict(row))
    conn.close()
    return result


@router.delete("/{college_id}")
def delete_college(college_id: int):
    conn = get_conn()
    if not conn.execute("SELECT id FROM colleges WHERE id = ?", (college_id,)).fetchone():
        conn.close()
        raise HTTPException(status_code=404, detail="College not found")
    conn.execute("DELETE FROM college_courses WHERE college_id = ?", (college_id,))
    conn.execute("DELETE FROM college_branches WHERE college_id = ?", (college_id,))
    conn.execute("DELETE FROM college_facilities WHERE college_id = ?", (college_id,))
    conn.execute("DELETE FROM colleges WHERE id = ?", (college_id,))
    conn.commit()
    conn.close()
    return {"message": "College deleted successfully"}

File: backend/routes/colleges.js
const express = require('express');
const router = express.Router();
const db = require('../db');

// Helper: attach courses, branches, facilities to a college row
function enrichCollege(college) {
  college.courses = db.prepare('SELECT course FROM college_courses WHERE college_id = ?')
    .all(college.id).map(r => r.course);
  college.branches = db.prepare('SELECT branch FROM college_branches WHERE college_id = ?')
    .all(college.id).map(r => r.branch);
  college.facilities = db.prepare('SELECT facility FROM college_facilities WHERE college_id = ?')
    .all(college.id).map(r => r.facility);
  return college;
}

// GET /api/colleges - list all with optional filters
router.get('/', (req, res) => {
  const { city, type, category, maxFees, search } = req.query;

  let query = 'SELECT * FROM colleges WHERE 1=1';
  const params = [];

  if (city) { query += ' AND city = ?'; params.push(city); }
  if (type) { query += ' AND type = ?'; params.push(type); }
  if (category) { query += ' AND category = ?'; params.push(category); }
  if (maxFees) { query += ' AND fees <= ?'; params.push(parseInt(maxFees)); }
  if (search) {
    query += ' AND (name LIKE ? OR city LIKE ? OR description LIKE ?)';
    const like = `%${search}%`;
    params.push(like, like, like);
  }

  const colleges = db.prepare(query).all(...params).map(enrichCollege);
  res.json(colleges);
});

// GET /api/colleges/:id - single college detail
router.get('/:id', (req, res) => {
  const college = db.prepare('SELECT * FROM colleges WHERE id = ?').get(req.params.id);
  if (!college) return res.status(404).json({ error: 'College not found' });
  res.json(enrichCollege(college));
});

// POST /api/colleges - add a new college
router.post('/', (req, res) => {
  const { name, city, type, category, fees, placement, avg_package, highest_package, seats, criteria, description, courses = [], branches = [], facilities = [] } = req.body;

  if (!name || !city || !type || !category || !fees) {
    return res.status(400).json({ error: 'name, city, type, category, fees are required' });
  }

  const insert = db.transaction(() => {
    const { lastInsertRowid: cid } = db.prepare(`
      INSERT INTO colleges (name, city, type, category, fees, placement, avg_package, highest_package, seats, criteria, description)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).run(name, city, type, category, fees, placement, avg_package, highest_package, seats, criteria, description);

    courses.forEach(c => db.prepare('INSERT INTO college_courses (college_id, course) VALUES (?, ?)').run(cid, c));
    branches.forEach(b => db.prepare('INSERT INTO college_branches (college_id, branch) VALUES (?, ?)').run(cid, b));
    facilities.forEach(f => db.prepare('INSERT INTO college_facilities (college_id, facility) VALUES (?, ?)').run(cid, f));

    return db.prepare('SELECT * FROM colleges WHERE id = ?').get(cid);
  });

  const college = insert();
  res.status(201).json(enrichCollege(college));
});

// PUT /api/colleges/:id - update a college
router.put('/:id', (req, res) => {
  const existing = db.prepare('SELECT * FROM colleges WHERE id = ?').get(req.params.id);
  if (!existing) return res.status(404).json({ error: 'College not found' });

  const { name, city, type, category, fees, placement, avg_package, highest_package, seats, criteria, description, courses, branches, facilities } = req.body;

  const update = db.transaction(() => {
    db.prepare(`
      UPDATE colleges SET
        name = COALESCE(?, name), city = COALESCE(?, city), type = COALESCE(?, type),
        category = COALESCE(?, category), fees = COALESCE(?, fees),
        placement = COALESCE(?, placement), avg_package = COALESCE(?, avg_package),
        highest_package = COALESCE(?, highest_package), seats = COALESCE(?, seats),
        criteria = COALESCE(?, criteria), description = COALESCE(?, description)
      WHERE id = ?
    `).run(name, city, type, category, fees, placement, avg_package, highest_package, seats, criteria, description, req.params.id);

    if (courses) {
      db.prepare('DELETE FROM college_courses WHERE college_id = ?').run(req.params.id);
      courses.forEach(c => db.prepare('INSERT INTO college_courses (college_id, course) VALUES (?, ?)').run(req.params.id, c));
    }
    if (branches) {
      db.prepare('DELETE FROM college_branches WHERE college_id = ?').run(req.params.id);
      branches.forEach(b => db.prepare('INSERT INTO college_branches (college_id, branch) VALUES (?, ?)').run(req.params.id, b));
    }
    if (facilities) {
      db.prepare('DELETE FROM college_facilities WHERE college_id = ?').run(req.params.id);
      facilities.forEach(f => db.prepare('INSERT INTO college_facilities (college_id, facility) VALUES (?, ?)').run(req.params.id, f));
    }

    return db.prepare('SELECT * FROM colleges WHERE id = ?').get(req.params.id);
  });

  res.json(enrichCollege(update()));
});

// DELETE /api/colleges/:id
router.delete('/:id', (req, res) => {
  const existing = db.prepare('SELECT * FROM colleges WHERE id = ?').get(req.params.id);
  if (!existing) return res.status(404).json({ error: 'College not found' });

  db.transaction(() => {
    db.prepare('DELETE FROM college_courses WHERE college_id = ?').run(req.params.id);
    db.prepare('DELETE FROM college_branches WHERE college_id = ?').run(req.params.id);
    db.prepare('DELETE FROM college_facilities WHERE college_id = ?').run(req.params.id);
    db.prepare('DELETE FROM colleges WHERE id = ?').run(req.params.id);
  })();

  res.json({ message: 'College deleted successfully' });
});

module.exports = router;
